### PR TITLE
Phase 1 Pixel DQM

### PR DESCRIPTION
**Pull request by schneiml**
This is a test/discussion PR. I am not sure whether it can/should/will be merged. @fioriNTU FYI.

This introduces the shared code for the new, geometry-independent Pixel DQM (atm called Phase1, though it should be very capable of also handling the old detector. It might even handle Strips, but this is all out of scope for now). This is still very rough and not fully implemented, but the supplied sample analyzer already produces some plots on digis.

This code relies on Max Stark's ( @DirtyDan88 ) work in the Alignment package. The commits are included in this PR, but they should probably be in a separate one. However, I could not find any. @mmusich @tlampen are there any plans on merging this stuff?

To make matters worse, I had to apply some changes to the Alignment package in commit https://github.com/cms-sw/cmssw/commit/30542c0547901805832fa3ae0de56adce105765e . These are untested, and might break things.

On more general things:
-  I don't know yet how the new could could and should be integrated in tests. Atm there are no automated tests.
- The code liberally uses `assert` to ensure it is used correctly. These should all be situations that only appear due to coding errors, but I still wonder whether exceptions are preferred in this case (backtraces might be useful).
- The shared code does not introduce a namespace. I think it could be a good idea, but I could not deduce the typical rules for when to use one (and how to call it).

### CODE REVIEW

**Comment by mmusich** [comment]
@schneiml we definitely plan to merge Max stuff (needed for next year alignment), but why can't that go separately? If you have dependencies can we build them up in a staggered approach?

**Comment by mmusich** [comment]
@smuzaffar the new packages should have DQM signature
@vanbesien @deguio

File: SLHCUpgradeSimulations/Configuration/python/phase1TkCustoms.py [hunk 1]
```diff
@@ -143,7 +143,9 @@ def customise_DQM(process,pileup):
     process.DQMOfflineTracking.remove(process.TrackingDQMSourceTier0Common)
 
     # Pixel DQM needs to be updated for phase1
-    process.DQMOfflinePreDPG.remove(process.siPixelOfflineDQM_source)
+    old = process.siPixelOfflineDQM_source
+    process.load('DQM.SiPixelPhase1Common.SiPixelPhase1OfflineDQM_source_cff')
+    process.DQMOfflinePreDPG.replace(old, process.siPixelPhase1OfflineDQM_source)
```
Comment (makortel) [inline SLHCUpgradeSimulations/Configuration/python/phase1TkCustoms.py]: @schneiml At this point please don't add new customizations to `phase1TkCustoms` (it's being cleaned up). Use the era mechanism instead, and delete the line removing `siPixelOfflineDQM_source` from the `DQMOfflinePreDPG` sequence.

Essentially you probably want to use `eras.phase1Pixel.toReplaceWith()` in https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/DQM/SiPixelCommon/python/SiPixelOfflineDQM_source_cff.py to replace the contents of `siPixelOfflineDQMSource`. See e.g. https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/RecoTracker/IterativeTracking/python/iterativeTk_cff.py or https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py for practical examples (in tracking configuration).

Said that, for me these changes are fine for testing or RFC-PRs (like this one seems to be at the moment), but for integration an era-based solution is needed.

File: SLHCUpgradeSimulations/Configuration/python/phase1TkCustoms.py [hunk 2]
```diff
@@ -259,6 +261,16 @@ def customise_harvesting(process):
 
     # Tracking DQM needs to be migrated for phase1
     process.DQMHarvestTracking.remove(process.TrackingOfflineDQMClient)
+    # Pixel DQM needs to be updated for phase1
+    #old = process.PixelOfflineDQMClientNoDataCertification
+    process.load('DQM.SiPixelPhase1Common.SiPixelPhase1OfflineDQM_harvesting_cff')
+    process.PixelOfflineDQMClientNoDataCertification.insert(1, process.siPixelPhase1OfflineDQM_harvesting)
```
Comment (makortel) [inline SLHCUpgradeSimulations/Configuration/python/phase1TkCustoms.py]: @schneiml Also this customization should eventually be done with eras.

**Comment by schneiml** [comment]
@mmusich I think we don't understand each other correctly. I would like to have the alignment stuff in a dedicated PR, but I don't feel I am the right person to open one. Esp. regarding the changes in https://github.com/cms-sw/cmssw/commit/30542c0547901805832fa3ae0de56adce105765e . 

And my code now depends on Max's code, so I cannot open a PR w/o these commits somewhere. I guess the right approach is doing a new PR for the Alignment code, get it merged, and then I rebase/reopen this PR on top.

@makortel Thanks for the hints. The entire configuration stuff in this PR is suboptimal, and I don't understand enough yet to fix it. I'll look into it.

**Comment by mmusich** [comment]
@schneiml thanks for the clarification. 
Can @DirtyDan88 or any Tracker alignment group representative open a PR with his commits + https://github.com/cms-sw/cmssw/commit/30542c0547901805832fa3ae0de56adce105765e  ?

**Comment by mmusich** [comment]
@mschrode @ghellwig @tlampen see above comment

**Comment by smuzaffar** [comment]
@schneiml, there is cyclic dependency b/w DQM/SiPixelPhase1Common & DQM/SiPixelPhase1Digis
(BuildFile and python cfi files). Do you really want to have SiPixelPhase1Digis with one plugin? I would suggest that move it in DQM/SiPixelPhase1Common/plugins directory

**Comment by schneiml** [comment]
@smuzaffar  the same sort of dependency exists in the old `DQM/SiPixelCommon`. In the end, there should be more `SiPixelPhase1*` packages like the digi one, most likely also some that have more than one plugin. 

This sounds reasonable to me, but the current SiPixel DQM is not necessarily known for good design. Would you recommend having the full Pixel DQM in one package? It should be much less code than the old one. I like the idea, it could reduce the amount of boilerplate to add a new plugin...

**Comment by smuzaffar** [comment]
If there are more SiPixelPhase1* like packages to come then I would suggest to design the packages structure now.

I do not mind having one DQM Pixal package as long as their external dependency is in control e.g. plugin with same set of dependency could live together.

**Comment by mmusich** [comment]
@ghellwig I think we agreed a new PR with the alignment only commits should have appeared early this week. is there any showstopper?

**Comment by ghellwig** [comment]
We are currently running a check to test Multi-IOV and full module-level alignment.
While doing that we fixed another bug in the new alignment framework.
We hope to get the comparison of old and new framework for the current geometry done within this week.
Btw. we had also fixed the same thing as the commit https://github.com/cms-sw/cmssw/commit/30542c0547901805832fa3ae0de56adce105765e, which @schneiml referenced.

**Comment by ghellwig** [comment]
@mmusich, @schneiml 
We have created a PR (#13976) with Max changes, also including your fix in https://github.com/cms-sw/cmssw/commit/30542c0547901805832fa3ae0de56adce105765e

Note, that it has been rebased to the tip of official-cmssw/CMSSW_8_0_X, so the commit IDs have changed.

**Comment by mmusich** [comment]
@ghellwig thanks, let's get the alignment stuff in first then.

**Comment by mmusich** [comment]
@schneiml 
https://github.com/cms-sw/cmssw/pull/13976 has been merged in the meanwhile. 
Do we still need this one? The branch doesn't merge anymore

**Comment by schneiml** [comment]
@mmusich no, this PR can be closed, #14230 is the current one.

Also, the code interfacing to Alignment is still present but `#if 0`'ed in the new PR since we now use our own naming convention. It might still be useful for testing now and then, but in the normal case the new PR is independent of alignment.

**Pull request closed without merging**